### ISSUE TITLE
checkout button is not responsive a few seconds after the button appears

### ISSUE THREAD

**Issue by MaxDBN** [paypal/paypal-checkout#471]
### Description

I get the warning ppxo_button_pre_template_click  and the button is not responsive when I click it , for a few seconds after the button appears

- [x] Screenshots or videos that show the issue
![paypalbug2](https://user-images.githubusercontent.com/16654524/30610626-72685af8-9d87-11e7-8a6e-1d69baa80f9d.png)

- [x] Your code
```
    CREATE_PAYMENT_URL  = 'http://localhost:3000/paypal/create_payment';
    EXECUTE_PAYMENT_URL = 'http://localhost:3000/paypal/execute_payment';

    paypal.Button.render({
      locale: 'en_US',
      env: 'sandbox',

      payment: function() {
        return new paypal.Promise(function(resolve, reject) {
          $('.paypal-overlay').fadeIn();
          $.post(CREATE_PAYMENT_URL)
            .done(function(data) { resolve(data.paymentID); })
            .fail(function(err)  { $('.paypal-overlay').fadeOut(); });
        });
      },

      onAuthorize: function(data) {
          console.log('The payment was authorized!');
          console.log('Payment ID = ',   data.paymentID);
          console.log('PayerID = ', data.payerID);

          $.post(EXECUTE_PAYMENT_URL, { paymentID: data.paymentID, payerID: data.payerID })
            .done(function(data) {
              window.location = '/payments/summary'
            })
            .fail(function(err)  {
              window.location = '/payments/cancelled'
            });
      },

      onCancel: function(data) {
        window.location = '/payments/cancelled'
      }

    }, '#paypal-buynow');
```
- [x] The `paypal.version` from your browser console
4.0.130
- [x] All console logs during the time of the issue, especially error messages

> logger.js:65 ppxo_button_pre_template_click {timestamp: 1505848188734, windowID: "2bda198901", pageID: "8143451158", referer: "developer.paypal.com", host: "developer.paypal.com", …}
> print @ logger.js:65
> log @ logger.js:175
> warn @ logger.js:224
> (anonymous) @ component.jsx:91

- [x] Does the issue also occur at [developer.paypal.com/demo/checkout](https://developer.paypal.com/demo/checkout)?
Yes

### Steps to reproduce

1. Refresh page
2. Rapidly click on checkout

### Affected browsers

- [x] Chrome
- [x] Firefox

**Comment by bluepnume** [comment]
Thanks for the report @MaxDBN. We actually do a pre-render of the button while the final iframe loads. We have plans to optimize this and allow clicks to the pre-rendered button, I'll keep this open until we do.

For now, at least, testing with `env: 'production'` should be faster -- from the statistics we have, loading the iframe should only take 2-300ms in production mode.

**Comment by Dryzen** [comment]
Hello, i just also found this issue in the local sandbox development and i don't like it.

**Comment by DanielLeone** [comment]
Just trying to understand what problem this is solving by showing a pre-render of the button?
Surely it'd be better to show a loading spinner, or even nothing at all, instead of a fake button that doesn't respond to clicks?

I feel like I'm abusing something by, "display: none"ing that button thing for a little while until it's actually ready.

**Comment by DanielLeone** [comment]
Trying to understand what problem this is solving by 'pre-rendering' the button.
Surely it'd be better to show a loading spinner, or even nothing at all, instead of a fake button which doesn't respond to clicks?

Maybe I'm misunderstanding something since I'm about to "display: none" this button thing for a little bit until it's actually ready. ¯\_(ツ)_/¯

**Comment by bluepnume** [comment]
@DanielLeone -- the rationale is, in production mode, the button should be visible to the user immediately (so they can see paypal is a payment option), and by the time the button is clicked it should have had enough time to load. From our logs we see this is the case the vast majority of the time. Can I ask -- are you in sandbox or production mode? Sandbox mode can be significantly slower, which makes this problem appear a lot worse than it actually is.

**Comment by jkomusin** [comment]
In addition to sharing @DanielLeone's confusion initially, we're seeing stylistic changes to the button between when it has "pre-rendered" to when it has "fully rendered":

Pre-render:
![screen shot 2017-12-06 at 11 30 15 am](https://user-images.githubusercontent.com/488939/33672789-15af44c8-da79-11e7-8abc-3e36210e6a69.png)

Fully-rendered:
![screen shot 2017-12-06 at 11 30 28 am](https://user-images.githubusercontent.com/488939/33672796-1b0cc382-da79-11e7-86fd-a7ec810a73c2.png)

I agree with your rationale that it is nice to provide an indication that PayPal is an available payment method immediately rather than waiting, but this results in an odd user experience. I like to assume that this is a mismatch between the version we have installed and the most current version PayPal is serving to us. If so, this is another unfortunate side-effect of this forceful pre-loading.

**Comment by bluepnume** [comment]
@jkomusin -- are you using the CDN version of the script? Or if not, which version are you using?

**Comment by jkomusin** [comment]
4.0.166, looks to be .001 behind NPM.

**Comment by bluepnume** [comment]
Are you using it from the evergreen paypalobjects/api/checkout.js url, or hard-coding the version?

**Comment by jkomusin** [comment]
We are installing from NPM and building it into our single-page app's bundle, so occasionally we will have a stale version until our app is re-built and deployed. Bumping the version to 4.0.167 (again, from NPM) results in a subtler transition, but the subtext still fades out after loading.

**Comment by bluepnume** [comment]
OK. I would strongly recommend using https://www.paypalobjects.com/api/checkout.js -- this will ensure there's no difference between the initial render and the iframe render. It will also give performance benefits.

**Comment by deliriousrhino** [comment]
Is there any update on this issue, I am using  https://www.paypalobjects.com/api/checkout.js and I still am able to click pay and nothing happened when the button first load

**Comment by alepore** [comment]
@bluepnume is there some class/attribute that we can check to know if the button is fully loaded or not?

**Comment by iamdevlinph** [comment]
Hi. Any update on this issue? I'm still experience the issue above where the button cannot be clicked right away.

I  noticed that the button can only be clicked once the request below is done.

```
Request URL: https://www.paypal.com/webapps/hermes/api/logger
Request Method: POST
```

**Comment by karan-levaral** [comment]
i am also facing this issue, this might decrease conversion ratio. I am using React button driver

**Comment by raRaRa** [comment]
Is there any temp workaround for this issue? E.g. checking if the iframe has loaded manually.

**Comment by raRaRa** [comment]
My final solution ended up looking like this:

The button itself, note that I'm assigning a ref to the button and I have the `onDisplay` event calling `this.onDisplayPP`:

```
<div className={classNames(this.state.ppButtonLoaded === false && 'hide')}>
    <PayPalButton
        ref={this.ppButton}
        commit={true}
        env="sandbox"
        client={{
            sandbox: '',
            production: '',
        }}
        payment={(data, actions) => this.payment(data, actions)}
        onAuthorize={this.onAuthorize}
        onDisplay={this.onDisplayPP}
    />
</div>
```

And here's the `onDisplayPP` function:

```
onDisplayPP = (e) => {
    if (this.ppButton.current && this.ppButton.current.state) {
        this.ppInterval = setInterval(() => {
            if (this.ppButton.current.state.parent.prerenderIframe.className.indexOf('xcomponent-invisible') > -1) {
                clearTimeout(this.ppInterval)
                this.setState({
                    ppButtonLoaded: true,
                })
            }
        }, 300) 
    }
}
```

The state.ppButtonLoaded will determine if the iframe has loaded. You can display anything while it's loading, such as "Loading PayPal", etc.

Hope this helps anyone else, and PLEASE note that this could break at any time.

**Comment by godhand1607** [comment]
It seems that the `onEnter` event is more reliable

**Comment by raRaRa** [comment]
@godhand1607 Thanks, I wasn't aware of that event! Works a lot better, since `onEnter` is called when the invisible class is removed. Perfect.

**Comment by bluepnume** [comment]
Closing this; this issue is fixed in the new version of the sdk which is currently in beta. Will update here when it's available for general release.

**Comment by bluepnume** [comment]
Yep, new major version which will require a manual upgrade. Hoping to go public beta by the end of the month.

**Comment by martinmckenna** [comment]
Adding my name to the people who need this feature. Not really sure why this issue is marked as closed when the release with the fix hasn't went live yet.

**Comment by bluepnume** [comment]
Please see:

https://developer.paypal.com/docs/checkout/
https://developer.paypal.com/docs/checkout/reference/upgrade-integration/

The latest script renders a button which is clickable from the moment it is rendered.

**Comment by martinmckenna** [comment]
Just to be clear, if I'm using `https://www.paypalobjects.com/api/checkout.v4.js` it should be as simple as using v5 now?

**Comment by bluepnume** [comment]
Not quite. I'd recommend taking a look at the upgrade guide.

**Comment by martinmckenna** [comment]
@bluepnume thanks for that. There's still one issue I can't figure out. I'm using the script in my React app and have the following in my `render` function:

```js
              <PaypalButton
                env={env}
                client={client}
                payment={this.payment} // not functional as of upgrade
                onApprove={this.onApprove}
                onCancel={this.onCancel}
                style={{
                  color: 'blue',
                  shape: 'rect'
                }}
              />
```

I can't seem to figure out the new handler for `payment.` It doesn't work after upgrading and I'm not able to see the upgrade path in the docs

**Comment by davidyeager** [comment]
This is a serious issue that could be hurting conversion depending on the delay people are experiencing.  For our setup the delay was around 2 seconds. The transition to the newer API is not clear when you're using a Braintree integration like ours. The issue was solved by initially rendering a button with the text "PayPal Loading..." and then changing that to the final button once enabled.

The key is to add `.then(onRender)` here: `paypal.Button.render(renderOptions, '#paypal-button').then(onRender);`

and then create some function `onRender()` that changes the button from your loading preview button to your final button. `paypal.Button.render.then()` seems to be called well after the button is rendered and exactly at the moment it is enabled.

**Comment by bluepnume** [comment]
@davidyeager as mentioned, the latest version of the sdk does not have this issue. Please upgrade if possible. https://developer.paypal.com/docs/checkout/reference/upgrade-integration/

**Comment by ccarstens** [comment]
@bluepnume As `@paypal/checkout-components` is not available via ES6 imports (#1109) can you please fix this bug for v4.x as well? 
 v4 (still available at npm as paypal-checkout) is referenced by your support team and in the documentation that the support team uses to outline how to implement express checkout?
https://braintree.github.io/braintree-web/3.34.0/PayPalCheckout.html
